### ISSUE TITLE
Add option to store output in a database

### ISSUE THREAD

**Issue by pouriyajamshidi** [pouriyajamshidi/tcping#46]
## Summary

We need to provide an option to store the output to a database preferably, `sqlite`.

This will further simplify integrating it to monitoring solutions such as `Prometheus` and `Grafana`

It'd be nice to have a wiki or a small tutorial showing how this can be integrated with the aforementioned monitoring solutions.

### Requirements

1. This [sqlite3 package](https://github.com/mattn/go-sqlite3) can be used to work with sqlite3 databases and store the data. I am not aware of any other ones but open to suggestions.

2. All the information that is being shown on the terminal, whether plane text or JSON (which is the same data in different formats) should be stored in a table with the name of the "target + start time" and have columns same as the terminal output. For instance check this output and pay attention to IP changes in the report as well:

    ![demo](https://github.com/pouriyajamshidi/tcping/raw/master/Images/gifs/tcping_resolve.gif)

3. The code has to go in a file called `db.go`
4. The flag should be `-d` or `-db`

P.S. This is a rough idea of how this should be implemented. Some requirements might change as progress is made.

**Comment by wizsk** [comment]
Hey @pouriyajamshidi, 
I would like to work on it. Can you please describe the it further?
What things should be saved in the db?

**Comment by pouriyajamshidi** [comment]
Hey @wizsk,

I have updated the issue with the requirements.

**Comment by wizsk** [comment]
How do you think about this schema?
```sql
CREATE TABLE PingResponses (
    id INTEGER PRIMARY KEY,
    destination_ip TEXT NOT NULL,
    port INTEGER NOT NULL,
    sequence_number INTEGER NOT NULL,
    reply_received BOOLEAN NOT NULL,
    reply_time_ms REAL,
    timestamp TIMESTAMP DEFAULT CURRENT_TIMESTAMP
);
```

**Comment by pouriyajamshidi** [comment]
Hi @wizsk 

`hostname` also has to come before `destination_ip`

I am not sure if I understand the difference between `reply_time_ms` and `timestamp`?

Also we need to make sure we are recording these too:

<img width="595" alt="image" src="https://github.com/pouriyajamshidi/tcping/assets/54482226/3968e171-3974-448b-86a4-0ea9c0f6e634">

**Comment by wizsk** [comment]
I'm not getting it..
but
My proposal is there should be a main table called `session`
in the session table other tables will be included?
or you will only have the one session and every time you se `tcping` the db will be reset?

**Comment by pouriyajamshidi** [comment]
Hi @wizsk,

The idea is to have a table with the name of the "target + start time"; for instance example_com_11_08_2023_17_00 and have columns same as the terminal output.

Wit this method, each target is in a separated table with a unique name and timestamp, making it easier to analyze later on.

I hope that clears it up

**Comment by wizsk** [comment]
hi @pouriyajamshidi 
I know I'm asking a lot of questions. I just want to do the work as the way you want it.

1. Should I create a new module called `db`? 
2. What do you thing about this schema? if you think something is not right please make changes. and submit a schema

I want to start working on it after your reply.

```sql
-- Table to store the master information (URL, port, and timestamp)
CREATE TABLE IF NOT EXISTS Master (
    id INTEGER PRIMARY KEY,
    url TEXT NOT NULL,
    port INTEGER NOT NULL,
    start_time DATETIME NOT NULL,
    end_time DATETIME NOT NULL
);

-- Table to store the detailed results for each probe
CREATE TABLE IF NOT EXISTS Results (
    id INTEGER PRIMARY KEY,
    master_id INTEGER NOT NULL,
    probe_number INTEGER NOT NULL,
    reply_received BOOLEAN NOT NULL,
    response_time FLOAT,
    FOREIGN KEY (master_id) REFERENCES Master (id)
);

-- Table to store the overall statistics
CREATE TABLE IF NOT EXISTS Statistics (
    id INTEGER PRIMARY KEY,
    master_id INTEGER NOT NULL,
    total_probes INTEGER NOT NULL,
    successful_probes INTEGER NOT NULL,
    unsuccessful_probes INTEGER NOT NULL,
    packet_loss_percentage FLOAT NOT NULL,
    total_uptime INTEGER NOT NULL,
    total_downtime INTEGER NOT NULL,
    longest_consecutive_uptime INTEGER NOT NULL,
    longest_consecutive_downtime INTEGER NOT NULL,
    rtt_min FLOAT NOT NULL,
    rtt_avg FLOAT NOT NULL,
    rtt_max FLOAT NOT NULL,
    start_time DATETIME NOT NULL,
    end_time DATETIME NOT NULL,
    FOREIGN KEY (master_id) REFERENCES Master (id)
);
```

**Comment by pouriyajamshidi** [comment]
Hi @wizsk,

> 1. Should I create a new module called db?

A new file named `db.go` and of course for its test, `db_test.go`

> 2. What do you thing about this schema? if you think something is not right please make changes. and submit a schema

I think we have a bit of a miscommunication here. The requirements are pretty simple and there is no need for multiple tables.

There should be **one** table per target. So if one runs `tcping -d example.com 443` at 11 AM on August 1st, 2023, a table should be created in the database with this name "example_com_443_11_00_00_01_08_2023" that contains columns for whatever you see on the terminal output when the program is done running. 

When the program is quit, say one presses `ctrl-c`, then the `stats` struct can be written to that table in the database.

Basically, you need to satisfy this interface: https://github.com/pouriyajamshidi/tcping/blob/d9d00f626d0695ee3a1daa98965e9da1109a35bc/tcping.go#L24

And this is where the magic happens: https://github.com/pouriyajamshidi/tcping/blob/d9d00f626d0695ee3a1daa98965e9da1109a35bc/tcping.go#L171


In terms of constructing the DB structure myself, I cannot do it now since my time is very limited but I hope I was able to clear the confusion.

**Comment by wizsk** [comment]
ok got it

**Comment by wizsk** [comment]
hi @pouriyajamshidi 
I have done some work and it's working now. I need to write some more tests though. Will you take a look?

**Comment by pouriyajamshidi** [comment]
> hi @pouriyajamshidi I have done some work and it's working now. I need to write some more tests and some comments though. Will you take a look? Here is the code https://github.com/wizsk/tcping/tree/add-option-for-saving-to-db or I can open a draft pr if you want

Hi @wizsk,

I saw your draft PR. once it is in ready state, I will start the review. This might take a while since I am traveling.
